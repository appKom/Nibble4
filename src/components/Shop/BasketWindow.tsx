import React, { FC } from "react";
import styled from "styled-components";

const BasketWindow: FC = () => {
  return <Container> </Container>;
};

export default BasketWindow;

const Container = styled.div`
  background-color: blue;
  width: 100%
  border: 1px solid black;
`;
