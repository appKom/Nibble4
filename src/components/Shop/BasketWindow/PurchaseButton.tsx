import React, { FC } from "react";
import styled from "styled-components";
import { OnlineBlue } from "style";

type ButtonProps = {
  text: string;
};

const PurchaseButton: FC<ButtonProps> = ({ text }: ButtonProps) => {
  return <Button> {text} </Button>;
};

export default PurchaseButton;

const Button = styled.button`
  background-color: ${OnlineBlue};
  font-size: 26px;
  color: white;
  border: none;
  border-radius: 3px;
  width: 80%;
  height: 90%;

  margin-left: auto;
  margin-right: auto;
`;
