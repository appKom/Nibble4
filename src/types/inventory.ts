export type Product = {
  pk: number;
  name: string;
  price: number;
  image: {
    sm: string;
  };
  category: {
    pk: number;
    name: string;
  };
};
