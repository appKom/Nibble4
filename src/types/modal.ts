export enum modalTypes {
  PURCHASE = "PURCHASE",
  COMPLETE = "COMPLETE",
  ERROR = "ERROR",
  DISABLED = "DISABLED",
  INACTIVE = "INACTIVE",
}
