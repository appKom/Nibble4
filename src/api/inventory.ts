import { Product } from "types/inventory";
import { get } from "./index";

const fetchInventory = async (url: string): Promise<Product[]> => {
  const response = await get({ url }).catch(() => Promise.resolve(undefined));
  if (response && response.ok) {
    const json = await response.json();
    return json.sort((a: Product, b: Product) => (a.name > b.name ? 1 : -1));
  }
  return [];
};

export default fetchInventory;
