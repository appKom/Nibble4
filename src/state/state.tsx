import React, { createContext, Dispatch, ReactNode, useReducer } from "react";
import {
  Product,
  CartItem,
  incrementCartItem,
  createCartItem,
  decrementCartItem,
} from "types/inventory";
import { User, OlCoinsUser } from "types/user";
import { modalTypes } from "types/modal";
import { Action, ActionTypes } from "state/actions";

type State = {
  user?: User;
  olCoinsUser?: OlCoinsUser;
  inventory: Product[];
  category: string;
  cart: { [id: number]: CartItem };
  modalState: modalTypes;
  newOlCoins: number;
};

const initialState: State = {
  olCoinsUser: { balance: 100 },
  inventory: [],
  category: "Alt",
  cart: {},
  modalState: modalTypes.DISABLED,
  newOlCoins: 0,
};

const getIncrementedCartItem = (id: number, state: State): CartItem => {
  if (state.cart[id]) {
    return incrementCartItem(state.cart[id]);
  }
  return createCartItem(id);
};

const getDecrementedCart = (id: number, state: State) => {
  if (state.cart[id].quantity > 1) {
    return { ...state.cart, [id]: decrementCartItem(state.cart[id]) };
  }
  const removedCart = { ...state.cart };
  delete removedCart[id];
  return { ...removedCart };
};

const updateNewOlCoins = (newValue: number, state: State) => {
  return state.newOlCoins + newValue;
};

const StateReducer = (state: State, action: Action): State => {
  switch (action.type) {
    case ActionTypes.SET_USER:
      return { ...state, user: action.payload };
    case ActionTypes.LOGOUT_USER:
      return {
        ...state,
        user: undefined,
        cart: {},
        category: "Alt",
        modalState: modalTypes.DISABLED,
      };
    case ActionTypes.SET_INVENTORY:
      return { ...state, inventory: action.payload };
    case ActionTypes.SET_CATEGORY:
      return { ...state, category: action.payload };
    case ActionTypes.SET_MODAL_STATE:
      return { ...state, modalState: action.payload };
    case ActionTypes.CHANGE_NEWOLCOINS:
      return {
        ...state,
        newOlCoins: updateNewOlCoins(action.payload, state),
      };
    case ActionTypes.ADD_TO_CART:
      return {
        ...state,
        cart: {
          ...state.cart,
          [action.payload]: getIncrementedCartItem(action.payload, state),
        },
      };
    case ActionTypes.REMOVE_FROM_CART: {
      return {
        ...state,
        cart: getDecrementedCart(action.payload, state),
      };
    }
    case ActionTypes.EMPTY_CART:
      return { ...state, cart: {} };
    default:
      return { ...state };
  }
};

type StateContextProps = {
  state: State;
  dispatch: Dispatch<Action>;
};

const initialContext: StateContextProps = {
  state: initialState,
  dispatch: () => null,
};

export const StateContext = createContext<StateContextProps>(initialContext);

type StateProviderProps = {
  children: ReactNode;
};

export const StateProvider = (props: StateProviderProps): JSX.Element => {
  const [state, dispatch] = useReducer(StateReducer, initialState);
  const { children } = props;

  return (
    <StateContext.Provider value={{ state, dispatch }}>
      {children}
    </StateContext.Provider>
  );
};
