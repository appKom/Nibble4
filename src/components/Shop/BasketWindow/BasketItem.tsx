import React, { useContext, FC } from "react";
import styled from "styled-components";
import { StateContext } from "state/state";

import { removeFromCart } from "state/actions";
import { deleteItemFromCart } from "state/actions";

import { Product } from "types/inventory";
import { IMAGE_URI } from "../../../api";
import QuantityInput from "atoms/QuantityInput";

type BasketItemProps = {
  id: number;
  quantity: number;
};

const BasketItem: FC<BasketItemProps> = ({ id, quantity }: BasketItemProps) => {
  const { state, dispatch } = useContext(StateContext);

  // const removeItem = () => dispatch(removeFromCart(id));

  const item: Product = state.inventory.find((e) => e.pk === id)!;
  const imageSrc = item.image ? IMAGE_URI(item.image.sm) : "";

  return (
    <Container
      onClick={() => {
        dispatch(removeFromCart(id));
      }}
    >
      <img
        src={
          imageSrc ? imageSrc : `${process.env.PUBLIC_URL}/images/noImage.png`
        }
        alt="Product image"
      />
      <div>
        <b> {item.name} </b>
      </div>
      <RedCrossWrapper>
        <RedCross
          onClick={() => {
            dispatch(deleteItemFromCart(id));
          }}
        >
          <path d="M15.854 12.854c-0-0-0-0-0-0l-4.854-4.854 4.854-4.854c0-0 0-0 0-0 0.052-0.052 0.090-0.113 0.114-0.178 0.066-0.178 0.028-0.386-0.114-0.529l-2.293-2.293c-0.143-0.143-0.351-0.181-0.529-0.114-0.065 0.024-0.126 0.062-0.178 0.114 0 0-0 0-0 0l-4.854 4.854-4.854-4.854c-0-0-0-0-0-0-0.052-0.052-0.113-0.090-0.178-0.114-0.178-0.066-0.386-0.029-0.529 0.114l-2.293 2.293c-0.143 0.143-0.181 0.351-0.114 0.529 0.024 0.065 0.062 0.126 0.114 0.178 0 0 0 0 0 0l4.854 4.854-4.854 4.854c-0 0-0 0-0 0-0.052 0.052-0.090 0.113-0.114 0.178-0.066 0.178-0.029 0.386 0.114 0.529l2.293 2.293c0.143 0.143 0.351 0.181 0.529 0.114 0.065-0.024 0.126-0.062 0.178-0.114 0-0 0-0 0-0l4.854-4.854 4.854 4.854c0 0 0 0 0 0 0.052 0.052 0.113 0.090 0.178 0.114 0.178 0.066 0.386 0.029 0.529-0.114l2.293-2.293c0.143-0.143 0.181-0.351 0.114-0.529-0.024-0.065-0.062-0.126-0.114-0.178z"></path>
        </RedCross>
      </RedCrossWrapper>
      <QuantityInput quantity={quantity} id={id}></QuantityInput>
      <PricetagWrapper>
        <b>{item.price}</b>kr
      </PricetagWrapper>
    </Container>
  );
};

export default BasketItem;

const RedCrossWrapper = styled.div`
  display: flex;
  justify-content: flex-end;
  align-items: end;
`;

const RedCross = styled.svg`
  position: relative; 
  display: block; 
  fill: red;
  height: 1.5em;
  width: 1.5em;
  xmlns: http://www.w3.org/2000/svg;
  background-color: pink;  
`;

const PricetagWrapper = styled.div`
  display: flex;
  justify-content: flex-end;
  align-item: flex-end;
  padding-top: 15px;
`;

const Container = styled.div`
  width: 90%;
  height: 15%;
  margin-left: auto;
  margin-right: auto;
  margin-top: 5px;
  margin-bottom: 15px;
  padding: 10px;
  box-shadow: 2px 2px 7px #888888;
  border-radius: 3px;

  display: grid;
  grid-template-rows: 1fr 1fr;
  grid-template-columns: 1fr 4fr 1fr;

  img {
    width: 75%;
    grid-row: 1 / span 2;
  }
`;
