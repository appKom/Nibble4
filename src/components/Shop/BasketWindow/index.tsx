import React, { useContext, FC, useState } from "react";
import styled from "styled-components";
import BasketItem from "./BasketItem";
import { OnlineOrange } from "utility/style";
import { StateContext } from "state/state";
import { calculateCartTotal } from "types/inventory";
import Button from "../../../atoms/Button";
import { setModal } from "state/actions";
import Modal from "components/Shop/Modal";
import { modalTypes } from "types/modal";

const BasketWindow: FC = () => {
  const { state, dispatch } = useContext(StateContext);
  const { cart, user, inventory, purchaseComplete } = state;

  const totalPrice = calculateCartTotal(cart, inventory);
  const insufficient = user!.balance - totalPrice <= 0 ? true : false;

  const ShowModal = (show: boolean) => dispatch(setModal(show));
  /*
  const purchase = () => {
    if (insufficient || totalPrice <= 0) return undefined;
    else {
      return ShowModal(true);
    }
  };
*/
  const purchase = () => {
    return ShowModal(true);
  };

  const basketItems = Object.keys(cart).map((key: string) => (
    <BasketItem
      key={key}
      id={Number(key)}
      quantity={cart[Number(key)].quantity}
    />
  ));

  return (
    <Container>
      <h2> Your Cart</h2>
      <ItemDiv>{basketItems}</ItemDiv>

      <CostDiv>
        <span>
          <b> Total Cost</b>
        </span>
        <span id="pris">
          <b> {totalPrice}kr</b>
        </span>
        <span id="olcoins">
          <b> 0Ã¸c</b>
        </span>
      </CostDiv>
      <Button onClick={purchase}>
        {insufficient ? "Insufficient" : "Purchase"}
      </Button>

      <Modal
        type={purchaseComplete ? modalTypes.COMPLETE : modalTypes.PURCHASE}
      />
    </Container>
  );
};

export default BasketWindow;

const Container = styled.div`
  display: grid;
  grid-template-rows: 1fr 8fr 1fr 1fr;

  text-align: left;
  background-color: white;
  width: 85%;
  height: 70%;
  border-top: 10px solid ${OnlineOrange};
  border-radius: 3px;
  box-shadow: 2px 2px 7px #888888;
  padding: 5px;
  margin-left: auto;
  margin-right: auto;
`;

const ItemDiv = styled.div`
  ::-webkit-scrollbar {
    width: 8px;
  }
  ::-webkit-scrollbar-thumb {
    background: ${OnlineOrange};
    border-radius: 2px;
  }

  overflow: auto;
`;

const CostDiv = styled.div`
  display: grid;
  grid-template-columns: 5fr 1fr;
  grid-template-rows: 1fr 1fr;

  margin-left: auto;
  margin-right: auto;
  margin-bottom: 10px;
  margin-top: 10px;
  width: 90%;
  text-align: left;
  padding: 5px;
  font-size: 16px;
  border-bottom: 1px dashed gray;
  border-top: 1px dashed gray;
  #olcoins {
    grid-column: 2;
    grid-row: 2;
  }
`;
