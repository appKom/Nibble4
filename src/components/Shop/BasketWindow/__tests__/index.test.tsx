import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import BasketWindow from "../index";
import { User } from "types/user";
import { StateContext, State, StateContextProps } from "state/state";
import { CartItem, createCartItem } from "types/inventory";
import { modalTypes } from "types/modal";
import { Product } from "types/inventory";

const mockProduct: Product = {
  pk: 95,
  name: "Powerking",
  description: "Energidrikk",
  price: 10,
  image: {
    sm: "/media/images/responsive/sm/38688f77-21b6-4936-8ba6-c976538fb110.png",
  },
  category: {
    pk: 3,
    name: "Drikke",
  },
};

const mockUser: User = {
  pk: 1337,
  balance: 1337,
  first_name: "SuperAdmin",
};

const mockCartItem: CartItem = createCartItem(95);

const mockStateWithUser: State = {
  user: mockUser,
  inventory: [],
  category: "Alt",
  cart: {},
  modalState: modalTypes.DISABLED,
};

const mockState: State = {
  user: mockUser,
  inventory: [mockProduct],
  category: "Alt",
  cart: { 95: mockCartItem },
  modalState: modalTypes.DISABLED,
};

const userContext: StateContextProps = {
  state: mockStateWithUser,
  dispatch: () => null,
};

const fullContext: StateContextProps = {
  state: mockState,
  dispatch: () => null,
};

afterEach(cleanup);

describe("Unit-testing for BasketWindow", () => {
  it("Should match snapshot", () => {
    const { asFragment } = render(
      <StateContext.Provider value={userContext}>
        <BasketWindow />
      </StateContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("Integration test for BasketWindow with a BasketItem", () => {
  it("Should render with an BasketItem", () => {
    const { asFragment } = render(
      <StateContext.Provider value={fullContext}>
        <BasketWindow />
      </StateContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
