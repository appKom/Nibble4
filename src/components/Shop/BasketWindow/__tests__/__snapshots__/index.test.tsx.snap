// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test for BasketWindow with a BasketItem Should render with an BasketItem 1`] = `
<DocumentFragment>
  <div
    class="sc-fubCfw fvlDXG"
  >
    <h2>
       Handlekurven din
    </h2>
    <div
      class="sc-pFZIQ sSvdJ"
    >
      <div
        class="sc-bdfBwQ faJJuY"
      >
        <img
          alt="Product image"
          src="https://online.ntnu.no//media/images/responsive/sm/38688f77-21b6-4936-8ba6-c976538fb110.png"
        />
        <span>
          <b>
             Powerking 
          </b>
        </span>
        <span />
        <span>
          Quantity:
          <b>
            1
          </b>
        </span>
        <span>
          <b>
            10
          </b>
          kr
        </span>
      </div>
    </div>
    <div
      class="sc-jrAGrp hYvsZc"
    >
      <span>
        <b>
           Sum :
        </b>
      </span>
      <span
        id="pris"
      >
        <b>
           10kr
        </b>
      </span>
    </div>
    <button
      class="sc-gsTCUz dhzAWv"
    >
      Kjøp
    </button>
  </div>
</DocumentFragment>
`;

exports[`Unit-testing for BasketWindow Should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-fubCfw fvlDXG"
  >
    <h2>
       Handlekurven din
    </h2>
    <div
      class="sc-pFZIQ sSvdJ"
    />
    <div
      class="sc-jrAGrp hYvsZc"
    >
      <span>
        <b>
           Sum :
        </b>
      </span>
      <span
        id="pris"
      >
        <b>
           0kr
        </b>
      </span>
    </div>
    <button
      class="sc-gsTCUz dhzAWv"
    >
      Kjøp
    </button>
  </div>
</DocumentFragment>
`;
